src/gba_cell.rs

```diff
@@ -1,66 +1,79 @@
-//! ```
-//!
-//! And then you can use the [`read`](GbaCell::read) and
-//! [`write`](GbaCell::write) methods to interact with the data:
-//!
-//! ```
-//! # static THE_COLOR: GbaCell<Color> = GbaCell::new(Color::new());
-//! let old_color = THE_COLOR.read();
-//!
-//! THE_COLOR.write(Color::default());
-//! ```
+//! Provides the [`GbaCell`] type.
 
 use core::{
-  cell::UnsafeCell,
   fmt::Debug,
-  mem::{align_of, size_of},
   num::{NonZeroI16, NonZeroI32, NonZeroI8, NonZeroU16, NonZeroU32, NonZeroU8},
-  panic::RefUnwindSafe,
+  ptr::NonNull,
 };
 
-use crate::{
-  fixed::Fixed,
-  interrupts::IrqFn,
-  keys::{KeyControl, KeyInput},
-  video::Color,
-};
+use crate::{irq::IrqBits, keys::KeyInput, video::Color};
 
-/// A GBA-specific wrapper around Rust's [`UnsafeCell`](core::cell::UnsafeCell)
-/// type.
+/// Marker trait bound for the methods of [`GbaCell`].
 ///
-/// Supports any data type that implements the [`GbaCellSafe`] marker trait.
+/// When a type implements this trait it indicates that the type can be
+/// atomically loaded/stored using a single volatile access.
 ///
-/// ## Safety Logic
+/// ## Safety
+/// The type must fit in a single register, and have an alignment equal to its
+/// size. Generally that means it should be one of:
 ///
-/// * LLVM thinks that ARMv4T only supports atomic operations via special atomic
-///   support library functions. This is true for the "complex" atomic ops like
-///   "fetch-add", but for individual load or store ops this is overkill.
-/// * If you directly write an Acquire/load, Release/store, or a Relaxed op with
-///   an associated `compiler_fence`, then LLVM does generate correct code.
-///   However, it will have very sub-optimal performance. LLVM will generate
-///   calls to the mythical atomic support library, when it should just directly
-///   use an `ldr` or `str` instruction.
-/// * In response to this LLVM nonsense, the `GbaCell` type just uses inline
-///   assembly to perform all accesses to the contained data.
-/// * When LLVM sees inline assembly, it is forced to defensively act as if the
-///   inline assembly might have done *anything* legally possible using the
-///   pointer and value provided to the inline assembly. This includes that the
-///   inline assembly *might* call the atomic support library to access the
-///   pointer's data using an atomic load or store. So LLVM has to treat the
-///   inline assembly as an atomic sync point.
-/// * However, inside the inline asm block we actually just use the single load
-///   or store op that we wanted.
+/// * an 8, 16, or 32 bit integer
+/// * a function pointer
+/// * a data pointer to a sized type
+/// * an optional non-null pointer (to function or sized data)
+/// * a `repr(transparent)` newtype over one of the above
+pub unsafe trait GbaCellSafe: Copy {}
+// Note(Lokathor): The list here is not exhaustive, it's just all the stuff I
+// thought of at the time. Add more as necessary.
```

Comment:
`f32` too! Not natively supported by the GBA ofc, but software emulation is still occasionally useful.